[Cosmos] Fix strictness issues in ```helper.ts```

bzhang0 opened the pull request:
This is a PR for discussion on solving strictness issues in ```src/common/helper.ts``` in #12745. After looking through the build errors, I noticed that ```helper.ts``` doesn't import anything that isn't strict. 

The main errors right now are listed below:
https://github.com/Azure/azure-sdk-for-js/blob/69026d943baba56e2bbe69e1ab49b906e5b7f570/sdk/cosmosdb/cosmos/log.txt#L480-L483

https://github.com/Azure/azure-sdk-for-js/blob/69026d943baba56e2bbe69e1ab49b906e5b7f570/sdk/cosmosdb/cosmos/src/common/helper.ts#L29-L35

https://github.com/Azure/azure-sdk-for-js/blob/69026d943baba56e2bbe69e1ab49b906e5b7f570/sdk/cosmosdb/cosmos/src/common/helper.ts#L274-L277

https://github.com/Azure/azure-sdk-for-js/blob/69026d943baba56e2bbe69e1ab49b906e5b7f570/sdk/cosmosdb/cosmos/src/common/helper.ts#L282-L285

@deyaaeldeen, I'm not quite sure how to approach this section. Do you have any ideas on how I can return a null/undefined value that won't violate ```strict```? I attempted returning ```null```, but I received the same error. 

Also @jeremymeng, as requested 😄.

bzhang0 commented:
On a side note, I've kept strict on true just for the purposes of seeing the errors. I've also turned off --pretty in the rushx scripts so the output is more clear when I view the log.txt file.

deyaaeldeen commented on sdk/cosmosdb/cosmos/tsconfig.json:
```diff
@@ -2,7 +2,7 @@
   "compilerOptions": {
     "module": "esnext",
-    "strict": false,
+    "strict": true,
```
could you remove the helpers file from the list in `tsconfig.strict.json` instead?

bzhang0 replied:
yes! I can do this

deyaaeldeen commented:
> @deyaaeldeen, I'm not quite sure how to approach this section. Do you have any ideas on how I can return a null/undefined value that won't violate strict? I attempted returning null, but I received the same error.

@bzhang0 my suggestion is to keep the runtime behavior as is as much as possible and update the types to match it. For example for the first issue, the return type of that function is:
```ts
function parseLink(resourcePath: string): {
  type: ResourceType;
  objectBody: {
    id: string;
    self: string;
  };
}
```
and should be updated to (notice the `?` next to field names):
```ts
function parseLink(resourcePath: string): {
  type?: ResourceType;
  objectBody?: {
    id: string;
    self: string;
  };
}
```
To allow for `undefined` to be a valid value for this fields.

bzhang0 commented:
> > @deyaaeldeen, I'm not quite sure how to approach this section. Do you have any ideas on how I can return a null/undefined value that won't violate strict? I attempted returning null, but I received the same error.
> 
> @bzhang0 my suggestion is to keep the runtime behavior as is as much as possible and update the types to match it. For example for the first issue, the return type of that function is:
> 
> ```ts
> function parseLink(resourcePath: string): {
>   type: ResourceType;
>   objectBody: {
>     id: string;
>     self: string;
>   };
> }
> ```
> 
> and should be updated to (notice the `?` next to field names):
> 
> ```ts
> function parseLink(resourcePath: string): {
>   type?: ResourceType;
>   objectBody?: {
>     id: string;
>     self: string;
>   };
> }
> ```
> 
> To allow for `undefined` to be a valid value for these fields.

I see! This makes a lot of sense - thank you!

bzhang0 commented:
I've updated the lines accordingly, and removed ```helpers.ts``` from ```tsconfig.strict.json```. I also added ```string | null``` to the signature of ```export function getResourceIdFromPath(resourcePath: string): string | null {```. I'm not sure if this is allowed, but I did not receive any more build errors from ```helpers.ts``` after running ```rushx build```. Let me know if this works.

deyaaeldeen commented:
> I've updated the lines accordingly, and removed helpers.ts from tsconfig.strict.json. I also added string | null to the signature of

> I'm not sure if this is allowed, but I did not receive any more build errors from helpers.ts after running rushx build. Let me know if this works.

@bzhang0 the changes look good to me. However, it looks like this module has transitive imports to others with strictness issues. In particular, it imports [CosmosClientOptions.ts](https://github.com/Azure/azure-sdk-for-js/blob/0fc1beec21d20f011f3d5c046adcc3c968eeddb0/sdk/cosmosdb/cosmos/src/CosmosClientOptions.ts) which in turn imports many modules with strictness issues. You can see these errors when you run `rushx check:src:strict` and here is a link to the errors in CI: https://dev.azure.com/azure-sdk/public/_build/results?buildId=1409102&view=logs&j=3dc8fd7e-4368-5a92-293e-d53cefc8c4b3&t=acc99708-be46-50ff-f197-4ecb2c3cab18&l=243.

This means we will not be able to remove helper.ts from the list in tsconfig.strict.json right now but we should merge your other changes.

For your next PR, I suggest picking a module that does not have even transitive imports to modules with strictness issues, this is a hard problem because you will basically need to traverse the graph that represents import edges until you find a node that does not have any outgoing import edge, if any! If there is no such module exists, then your best bet is to find cycles and break them into smaller modules where at least one will be a node with no outgoing import edges. In a comment in the github issue, I suggested using an npm package to help doing this analysis, did you try it?

bzhang0 commented:
@deyaaeldeen Apologies for the late response! School work has been catching up to me again 😅. I went back to your post and saw the [import-graph](https://www.npmjs.com/package/import-graph) tool, but I was unable to use `npm i --S import-graph` to install it. Would it be possible to see if it works on your end?

bzhang0 commented:
Additionally, are changes here okay to be merged? Let me know if there are any outstanding errors coming from helper.ts.

bzhang0 commented:
Actually, I realized I wasn't in the main directory when installing import-graph. I'll play around with the tool and let you know if I have any other questions :)

deyaaeldeen commented on sdk/cosmosdb/cosmos/tsconfig.strict.json:
```diff
@@ -12,7 +12,6 @@
     "src/documents/DatabaseAccount.ts",
     "src/documents/IndexingPolicy.ts",
     "src/plugins/Plugin.ts",
-    "src/common/helper.ts",
```
Please add this line back to get the CI to be green again.

deyaaeldeen reviewed (approved):
Looks good, thanks!
